### PR TITLE
[GPU] Optimize Conv and Reorder pattern optimize in Unet

### PR DESCRIPTION
**Pull request by hyunback**
### Details:
 - *MVN with rank size 3 always requires Reorder and Reshape. Due to this pattern, too many Reorder may occur when used with Convolution, which may performance degradation. It stand out in Stable-Diffusion Unet and Decoder*
 - *Convolution performance itself also has the gain in MTL.*

### Tickets:
 - *137360*

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 1]
```diff
@@ -685,6 +687,35 @@ bool layout_optimizer::convolution_b_fs_yx_fsv16_opt(const layout& input_layout,
...
+    if (cur_depth > max_depth) return false;
+    if (node.is_type<reorder>()) {
+        for (auto& reorder_user : node.get_users()) {
+            if (reorder_user->is_type<reshape>()) {
+                for (auto& reshape_user : reorder_user->get_users()) {
+                    if (reshape_user->is_type<mvn>()) {
+                        if (reorder_size_threshold == 0) {
+                            return true;
+                        } else {
+                            if (node.get_output_layout().get_linear_size() > reorder_size_threshold) {
```
Comment (e-ddykim) [inline src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:702]: Do we need the code at L699 - L701? It looks like that L702 is enough to check the zero threshold condition.

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 2]
```diff
@@ -685,6 +687,35 @@ bool layout_optimizer::convolution_b_fs_yx_fsv16_opt(const layout& input_layout,
     return false;
 }
 
+static bool has_rank3_mvn_user(const program_node& node, size_t cur_depth, size_t max_depth, uint64_t reorder_size_threshold = 0) {
```
Comment (e-ddykim) [inline src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:690]: Actually, this function `has_rank3_mvn_user` does not check the mvn's input rank. What do you think of renaming it as `has_reorder_before_mvn'?

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 3]
```diff
@@ -1202,7 +1213,9 @@ format layout_optimizer::get_expected_format(convolution_node const& node) {
                   || (_optimized_conv_count.at({format::fs_b_yx_fsv32, false}) * cond_denom > 0.8f)) &&
                   convolution_fs_b_yx_fsv32_opt(input_layout,
                                                 output_layout,
-                                                weights_layout, prim, true))))) {
+                                                weights_layout, prim, true)))) &&
+                 !(has_rank3_mvn_user(reinterpret_cast<program_node const&>(*node.get_users().front()), 0, 3, 1000000) &&
+                     !static_cast<bool>(prepare_padding::get_convolution_needed_padding(const_cast<convolution_node&>(node))))) {
```
Comment (e-ddykim) [inline src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:1218]: `node` is already of type `convolution_node`. Do we need the `const_cast`?

**Review by e-ddykim — APPROVED** [review]
It looks good to me except for some minor comments.

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp (hunk 2 above)
Comment (hyunback) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp]: Okay, applied.

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp (hunk 3 above)
Comment (hyunback) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:1212]: Need const_cast. node is `const convolution_node` in here, and `get_convolution_needed_padding` needs non const type. `get_output_layout()` used in `get_convolution_needed_padding`, has differently implemented const and non const API.  CI failure in unit-test with const case.

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 4]
```diff
@@ -685,6 +687,29 @@ bool layout_optimizer::convolution_b_fs_yx_fsv16_opt(const layout& input_layout,
     return false;
 }
 
+static bool has_reorder_before_mvn(const program_node& node, size_t cur_depth, size_t max_depth, uint64_t reorder_size_threshold = 0) {
+    // MVN with rank size 3 always requires Reorder and Reshape. Due to this pattern, too many Reorder may occur when used with Convolution,
+    // which may performance degradation. It stand out in Stable-Diffusion Unet and Decoder.
+    if (cur_depth > max_depth) return false;
+    if (node.is_type<reorder>()) {
+        for (auto& reorder_user : node.get_users()) {
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:695]: do we need to take multi-user scenario into consideration? I expect this is only for single-user case. If so, limiting the scope(for single-user only) will be easy to maintain..

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 5]
```diff
@@ -685,6 +687,29 @@ bool layout_optimizer::convolution_b_fs_yx_fsv16_opt(const layout& input_layout,
     return false;
 }
 
+static bool has_reorder_before_mvn(const program_node& node, size_t cur_depth, size_t max_depth, uint64_t reorder_size_threshold = 0) {
+    // MVN with rank size 3 always requires Reorder and Reshape. Due to this pattern, too many Reorder may occur when used with Convolution,
+    // which may performance degradation. It stand out in Stable-Diffusion Unet and Decoder.
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:692]: nit: `which may performance degradation` -> `which may cause performance degradation`
`It stand out` -> `It stands out`

File: src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_padding.cpp [hunk 6]
```diff
@@ -248,14 +194,80 @@ void prepare_padding::run(program& p) {
...
-        //    round_up_to(left_padding + prev_prim_output_layout.spatial(0) + right_padding, 16));
-        // right_padding = needed_buffer_size_x - left_padding - prev_prim_output_layout.spatial(0);
-
-        cldnn::padding needed_padding({0, 0, padding_begin_x, padding_begin_y, padding_begin_z}, {0, 0, padding_end_x, padding_end_y, padding_end_z}, 0);
-        needed_padding = padding::max(prev_prim_output_layout.data_padding, needed_padding);
         p.apply_needed_padding(node, node.get_dependency(0), needed_padding);
     }
 }
+
+cldnn::padding prepare_padding::get_convolution_needed_padding(convolution_node& node) {
```
[the remark this answers is not in the document]
Comment (hyunback) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/prepare_padding.cpp]: Applied

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp (hunk 4 above)
Comment (hyunback) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp]: Applied, I checked and no changes in kernel selection.

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp (hunk 5 above)
Comment (hyunback) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp]: Applied.

**Review by isanghao — APPROVED** [review]

**Pull request merged**